File: FactoryMethod/FactoryDemo.js
class Computer {
  description;

  getInfo(){
    return this.description;
  }
}

class PC extends Computer {
  makeComputer(cpu, ram, storage, display){
    this.description = `Dell PC: CPU ${cpu}, RAM ${ram}, Storage ${storage}, Display ${display}`
  }
}

class Laptop extends Computer {
  makeComputer(cpu, ram, storage, display){
    this.description = `Dell Laptop: CPU ${cpu}, RAM ${ram}, Storage ${storage}, Display ${display}`
  }
}

class OrderSystem {
  productList = []

  placeOrder(type, cpu, ram, storage, display){
    const product = this.create(type, cpu, ram, storage, display);

    if (product)
      this.productList.push(product)
    
  }


  listOrder(){
    this.productList.forEach(product => console.log(product.getInfo()))
  }


  create(type, cpu, ram, storage, display){}

}

class Order extends OrderSystem{
  creatorList = {
    'pc': () => new PC(),
    'laptop': () => new Laptop()
  }

  create(type, cpu, ram, storage, display){
    const creator = this.creatorList[type.toLowerCase()]?.();

    if (creator)
      creator.makeComputer(cpu, ram, storage, display)
    

    return creator
  }
}

const order = new Order();

pcConfig = {
  type: "PC",
  cpu: 4, 
  ram: 14, 
  storage: 334, 
  display: 15
}

order.placeOrder("PC", 4, 14, 334, 15);
order.placeOrder("Laptop", 1, 4, 512, 14)
order.listOrder()

File: FactoryMethod/FactoryDemo.java
// package Factory.Java.Problem;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;


abstract class Product {
  protected String description;

  public abstract void makeComputer(int cpu, int ram, int storage, int display);

  public String getInfo() {
    return this.description;
  }
}

class PC extends Product {
  public void makeComputer(int cpu, int ram, int storage, int display) {
    this.description = MessageFormat.format("DELL PC: CPU {0}, RAM {1}, STORAGE {2}, DISPLAY {3}", cpu, ram, storage,
        display);
  }
}

class Laptop extends Product {
  public void makeComputer(int cpu, int ram, int storage, int display) {
    this.description = MessageFormat.format("DELL Laptop: CPU {0}, RAM {1}, STORAGE {2}, DISPLAY {3}", cpu, ram,
        storage, display);
  }
}


public class FactoryDemo {

  public static Product create(String type, int cpu, int ram, int storage, int display) {
    Product product = null;
    switch (type) {
      case "PC": 
        product = new PC();
        product.makeComputer(cpu, ram, storage, display);
        break;
      case "Laptop":
        product = new Laptop();
        product.makeComputer(cpu, ram, storage, display);
        break;
    }
    return product;
  }

  public static void main(String[] args) {

    var computer = create("PC", 4, 32, 2, 16);

    System.out.println(computer.getInfo());
  }
}
